diff --git a/libavformat/file.c b/libavformat/file.c
index 264542a36a..8e8a02db4f 100644
--- a/libavformat/file.c
+++ b/libavformat/file.c
@@ -37,6 +37,10 @@
 #include <stdlib.h>
 #include "os_support.h"
 #include "url.h"
+#include "emscripten.h"
+
+extern int js_is_queue_empty(int queueId);
+extern int js_read_from_queue(int queueId, unsigned char *buf, int count);
 
 /* Some systems may not have S_ISFIFO */
 #ifndef S_ISFIFO
@@ -115,6 +119,26 @@ static int file_read(URLContext *h, unsigned char *buf, int size)
     return (ret == -1) ? AVERROR(errno) : ret;
 }
 
+static int pipe_read(URLContext *h, unsigned char *buf, int size)
+{
+    FileContext *c = h->priv_data;
+    int ret;
+
+    size = FFMIN(size, c->blocksize);
+
+    /* yield if the queue is empty */
+    while (js_is_queue_empty(c->fd)) {
+      emscripten_sleep(1);
+    }
+
+    ret = js_read_from_queue(c->fd, buf, size);
+
+    if (ret == 0 && c->follow) {
+        return AVERROR(EAGAIN);
+    }
+    return (ret == -1) ? AVERROR(errno) : ret;
+}
+
 static int file_write(URLContext *h, const unsigned char *buf, int size)
 {
     FileContext *c = h->priv_data;
@@ -386,7 +410,7 @@ static int pipe_open(URLContext *h, const char *filename, int flags)
 const URLProtocol ff_pipe_protocol = {
     .name                = "pipe",
     .url_open            = pipe_open,
-    .url_read            = file_read,
+    .url_read            = pipe_read,
     .url_write           = file_write,
     .url_get_file_handle = file_get_handle,
     .url_check           = file_check,
