diff --git a/libavformat/file.c b/libavformat/file.c
index 264542a36a..02cc77fb3c 100644
--- a/libavformat/file.c
+++ b/libavformat/file.c
@@ -37,6 +37,7 @@
 #include <stdlib.h>
 #include "os_support.h"
 #include "url.h"
+#include "emscripten.h"
 
 /* Some systems may not have S_ISFIFO */
 #ifndef S_ISFIFO
@@ -115,6 +116,36 @@ static int file_read(URLContext *h, unsigned char *buf, int size)
     return (ret == -1) ? AVERROR(errno) : ret;
 }
 
+static int pipe_read(URLContext *h, unsigned char *buf, int size)
+{
+    FileContext *c = h->priv_data;
+    int ret;
+
+#if __EMSCRIPTEN__
+    /* yield if the stdinQueue is empty */
+    if (EM_ASM_INT({
+          if (drain) return 0;
+          dropInput();
+          if (stdinQueue.length === 0) return 1;
+          return 0;
+        }, 0)) {
+      emscripten_sleep(1);
+    }
+#endif
+
+    size = FFMIN(size, c->blocksize);
+    ret = read(c->fd, buf, size);
+
+#if __EMSCRIPTEN__
+    EM_ASM_({ console.log('read ' + $0 + ' bytes'); }, ret);
+#endif
+
+    if (ret == 0 && c->follow) {
+        return AVERROR(EAGAIN);
+    }
+    return (ret == -1) ? AVERROR(errno) : ret;
+}
+
 static int file_write(URLContext *h, const unsigned char *buf, int size)
 {
     FileContext *c = h->priv_data;
@@ -386,7 +417,7 @@ static int pipe_open(URLContext *h, const char *filename, int flags)
 const URLProtocol ff_pipe_protocol = {
     .name                = "pipe",
     .url_open            = pipe_open,
-    .url_read            = file_read,
+    .url_read            = pipe_read,
     .url_write           = file_write,
     .url_get_file_handle = file_get_handle,
     .url_check           = file_check,
